### PR TITLE
QR Code stricter validation (EXPOSUREAPP-2570)

### PR DESCRIPTION
**Pull request by AndroidMedaGalaxy**
## Description

This PR implements the Stricter QR code validations 

Earlier only the presence of GUID was checked, now the URL format will be checked as well

### CODE REVIEW

**Review by chris-cwa — DISMISSED** [review]
I like it!

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionConstants.kt [hunk 1]
```diff
@@ -49,6 +49,8 @@ object SubmissionConstants {
 
     const val minKeyCountForSubmission = 14
     const val fakeKeySize = (1 * 16 /* key data*/) + (3 * 4 /* 3x int32*/)
+    const val qrCodeRegex = "^(http:\\/\\/www\\.|https:\\/\\/www\\.|http:\\/\\/|https:\\/\\/)" +
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionConstants.kt:52]: Please make constants `ALL_CAPS`.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt [hunk 2]
```diff
@@ -75,6 +76,11 @@ object SubmissionService {
         return !(potentialGUID.isEmpty() || potentialGUID.length > SubmissionConstants.MAX_GUID_LENGTH)
     }
 
+    private fun matchGUIDPattern(scanResult: String): Boolean {
+        val pattern = Regex(SubmissionConstants.qrCodeRegex)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt:80]: Don't recompile the regex each time, put `Pattern.compile(....)` into a constant and re-use it. Regex pattern compilation can be costly.

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/service/submission/SubmissionServiceTest.kt [hunk 3]
```diff
@@ -171,11 +171,17 @@ class SubmissionServiceTest {
...
-        // invalid
+        // invalid if no GUID
         assertThat(
             SubmissionService.containsValidGUID("https://no-guid-here"),
             equalTo(false)
         )
+
+        // invalid if url format is incorrect
+        assertThat(
+            SubmissionService.containsValidGUID("htps://wrongformat.com"),
```
Comment (d4rken) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/service/submission/SubmissionServiceTest.kt:182]: Can you use our new assertion framework here? I think it's already available in `dev`, try:

`SubmissionService.containsValidGUID("htps://wrongformat.com") shouldBe false`

**Review by d4rken — CHANGES_REQUESTED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionConstants.kt (hunk 1 above)
Comment (mlenkeit) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionConstants.kt:52]: This should be checking for https only and for `localhost`. Please also add corresponding unit tests for localhost and other host names.

**Review by mlenkeit — CHANGES_REQUESTED** [review]

**Review by mlenkeit — CHANGES_REQUESTED** [review]
The regexp should check for `https://localhost?GUID` (with a valid GUID being `3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA`). Anything else should be rejected. Please add corresponding unit tests to cover scenarios such as
- https://localhost/?3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA -> pass
- https://localhost/%20?3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA -> not pass
- https://some-host.com/?3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA -> not pass
- https://localhost/?3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA -> not pass
- https://localhost/?4CD1F87D6FDA -> not pass
anything else (e.g. existing test cases) should fail

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/service/submission/SubmissionServiceTest.kt (hunk 3 above)
Comment (ritsam) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/service/submission/SubmissionServiceTest.kt]: it's been done.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt (hunk 2 above)
Comment (ritsam) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt]: it's been done.  

note: not a const variable as "Pattern" is not Primitive data type

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionConstants.kt (hunk 1 above)
Comment (ritsam) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionConstants.kt]: @mlenkeit  the relevant regex has been added

I have included all scenarios in test cases but you're welcome to test yourself [here](https://regexr.com/)

Regex string : ^((^https:\/{2}localhost)(\/\?)[A-Fa-f0-9]{6}[-][A-Fa-f0-9]{8}[-][A-Fa-f0-9]{4}[-][A-Fa-f0-9]{4}[-][A-Fa-f0-9]{4}[-][A-Fa-f0-9]{12})$

Test String: https://localhost/?3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA

**Comment by ritsam** [comment]
> The regexp should check for `https://localhost?GUID` (with a valid GUID being `3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA`). Anything else should be rejected. Please add corresponding unit tests to cover scenarios such as
> 
> * https://localhost/?3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA -> pass
> * https://localhost/%20?3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA -> not pass
> * https://some-host.com/?3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA -> not pass
> * https://localhost/?3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA -> not pass
> * https://localhost/?4CD1F87D6FDA -> not pass
>   anything else (e.g. existing test cases) should fail

These tests have been added. Existing test cases  were failing, they're amended as well

**Review by d4rken — CHANGES_REQUESTED** [review]
Please check whether all the scanresult validation could be made part of a neat little data class, i.e.

```kotlin
data class ScanResult(val rawResult: String) {
    val isValid by lazy {
        // do checks here.
    }
    val guid: String? by lazy {
        if(isValid) ... else null
    }
   companion object {
     // related const stuff  
    }
}

fun validateAndStoreTestGUID(rawResult: String) {
        val scanResult = ScanResult(rawResult)
        if (scanResult.isValid) {
            SubmissionService.storeTestGUID(scanResult.guid)
            _scanStatus.value = Event(ScanStatus.SUCCESS)
        } else {
            _scanStatus.value = Event(ScanStatus.INVALID)
        }
    }

```

**Review by mlenkeit — DISMISSED** [review]
LGTM from a requirement's perspective 👍  But I think it's worth having a fellow Engineer reviewing the code, too.

**Comment by ritsam** [comment]
> ```kotlin
> SubmissionService.storeTestGUID(scanResult.guid)
> ```

@d4rken  Please check out the new logic. I have tried to make it easier and keep it simple at the same time. 
let me know what you think.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    val isValid by lazy {
+        containsValidGUID(rawResult)
+    }
+    val guid: String? by lazy {
+        if (isValid) guid
+        else null
+    }
+
+    private fun matchGUIDPattern(scanResult: String): Boolean {
+        val pattern = QR_CODE_REGEX
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt:15]: Could inline `pattern` here.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        return pattern.toRegex().matches(scanResult)
+    }
+
+    private fun containsValidGUID(rawResult: String): Boolean {
+        if (rawResult.length > SubmissionConstants.MAX_QR_CODE_LENGTH ||
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt:20]: Make all constants that are part of the whole `ScanResult` stuff, part of this class such that we have a contained solution.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+            rawResult.count { it == SubmissionConstants.GUID_SEPARATOR } != 1 ||
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt:21]: Same here, refactor into this class.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            !matchGUIDPattern(rawResult)
+        )
+            return false
+
+        val potentialGUID = SubmissionService.extractGUID(rawResult)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt:26]: Also move this into this class, all other instances where someone needs to call `extractGUID` they should work with the `ScanResult` class.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        return !(potentialGUID.isEmpty() || potentialGUID.length > SubmissionConstants.MAX_GUID_LENGTH)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/ScanResult.kt:27]: Into this class

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt [hunk 9]
```diff
@@ -65,20 +65,8 @@ object SubmissionService {
...
-
-        val potentialGUID = extractGUID(scanResult)
-
-        return !(potentialGUID.isEmpty() || potentialGUID.length > SubmissionConstants.MAX_GUID_LENGTH)
-    }
-
-    private fun matchGUIDPattern(scanResult: String): Boolean {
-        val pattern = SubmissionConstants.QR_CODE_REGEX
-        return pattern.toRegex().matches(scanResult)
+        val scanResult = ScanResult(scanResult)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt:68]: Close, but go even further, i.e. in `SubmissionViewModel.validateAndStoreTestGUID` start working with `ScanResult`.

The core idea is to stop passing `String` and start passing `ScanResult` as early as possible.

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt:68]: Once refactored, move all related scan result tests into their own `ScanResultTest` class.

**Review by d4rken — CHANGES_REQUESTED** [review]

**Review by d4rken — APPROVED** [review]
lgtm

**Pull request merged**